### DISCUSSION TITLE
Typescript problem: can't use actions inside always

### DISCUSSION THREAD

**Discussion by DirkWolthuis** [statelyai/xstate#1480] [General] [answered]
Hi all, 

With TS version 4.0.3 I get a Typescript error that I can't place.  

```typescript
 starting: {
        always: [
          { target: "loadingGenerator", cond: "checkId" },
          {
            target: "editGenerator",
            actions: assign({
              generator: () => new Generator(),
            }),
          },
        ],
      },
```

The actions object gives the following typescript error: `Type 'AssignAction<{ generator: Generator; }, GeneratorEvents>' is not assignable to type 'string | ActionObject<GeneratorContext, GeneratorEvents> | ActionFunction<GeneratorContext, GeneratorEvents> | Action<...>[] | undefined'.
  Type 'AssignAction<{ generator: Generator; }, GeneratorEvents>' is not assignable to type 'string'.ts(2322)`

The context is of type generator: Generator, so I don't think that is a problem.

**Comment by DirkWolthuis** [comment]
And the code works just fine :)

**Reply by davidkpiano** [reply]
Have you tried `assign<YourContext>({ ... })`?

**Comment by DirkWolthuis** [comment]
![image](https://user-images.githubusercontent.com/9962202/94288552-d17ba380-ff57-11ea-8c3e-a0fb17bfe704.png)
Still the same problem

**Reply by davidkpiano** [reply]
✏️ Please create a [CodeSandbox (XState TypeScript template)](https://codesandbox.io/s/xstate-typescript-template-6dsws) reproduction of this. Thanks!

**Reply by DirkWolthuis** [reply]
https://codesandbox.io/s/xstate-typeproblem-jndq5?file=/src/index.ts

Same code and typescript version but no error?

**Reply by Andarist** [reply]
How can you be sure that this is using the same TS version as you do? I can't find the TS version to be specified anywhere in this codesandbox.

**Comment by DirkWolthuis** [comment]
In the right bottom corner. Same place as VSC

**Reply by Andarist** [reply]
Ah, right - I could have sworn that I've looked there because I've remembered that it was working like VScode but didn't spot it in a rush 🤷‍♂️ 

So if everything works in the sandbox then the problem might be related to a configuration or some other project exclusive circumstances. Could you try to prepare a repository with the issue reproduced?

**Reply by DirkWolthuis** [reply]
My repo with the problem is public (https://gitlab.com/DirkWolthuis/dutch-dnd-tables). My TS version of VSC is set to 4.0.3 with is also the TS version in the project (located in packages/front). Sorry spending your time with this :(

**Reply by Andarist** [reply]
Once you provide any generic you have to provide them all - so in this case you can solve this by adding the second one, so it should be:
```js
actions: assign<GeneratorContext, GeneratorEvents>(/* ... */)
```

Ideally, this would work without any generics provided explicitly though. It's a known problem with our typings right now - I'm thinking actively how to fix this but it has proven to be hard. Our work on `xstate-codegen` should aim to solve this - but, of course, this will require an additional setup etc. The benefits will outweigh the costs though 🤞 

The problem arises because TS doesn't try to match the return type to the context type, it's somehow related to the return type being a Partial of our TContext generic. You can trick this by writing this:
```js
actions: assign((ctx) => {
  return {
    ...ctx,
    generator: {
      name: "",
      shared: true,
      description: "",
      language: "nl",
      user_id: 1,
    },
  }
})
```

**Comment by DirkWolthuis** [comment, marked as answer]
cool thanks that works!

>

**Reply by davidkpiano** [reply]
Reference to working answer:

> Once you provide any generic you have to provide them all - so in this case you can solve this by adding the second one, so it should be:
> 
> ```js
> actions: assign<GeneratorContext, GeneratorEvents>(/* ... */)
> ```
> 
> Ideally, this would work without any generics provided explicitly though. It's a known problem with our typings right now - I'm thinking actively how to fix this but it has proven to be hard. Our work on `xstate-codegen` should aim to solve this - but, of course, this will require an additional setup etc. The benefits will outweigh the costs though 🤞
> 
> The problem arises because TS doesn't try to match the return type to the context type, it's somehow related to the return type being a Partial of our TContext generic. You can trick this by writing this:
> 
> ```js
> actions: assign((ctx) => {
>   return {
>     ...ctx,
>     generator: {
>       name: "",
>       shared: true,
>       description: "",
>       language: "nl",
>       user_id: 1,
>     },
>   }
> })
> ```

https://github.com/davidkpiano/xstate/discussions/1480#discussioncomment-85487
